Issue#34 create user entity and configure auth

isvyry opened the pull request:
dev
## JIRA

* [Main JIRA ticket](https://jira.softserve.academy/secure/RapidBoard.jspa?rapidView=id)


## Code reviewers

- [ ] @alexorlenko
- [ ] @Arthurpanda
- [ ] @shiftko  
- [ ] @mykytam 
- [ ] @omelchenkotimur 
- [ ] @Prosperro-de  
- [ ] @reeztl 

### Second Level Review

- [ ] @DzigMS 
- [ ] @kh0ma 

## Summary of issue

We need our application to have a basic security aspect - Auth.

## Summary of change

Created initial User entity and configured basic authentication via spring security.

## Testing approach

 ----

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

mykytam committed:
Issue#34: Create User entity and configure Auth
Add spring security dependency
Add User model and Role enum
Add basic SecurityConfig file

isvyry committed:
Merge branch 'dev' into Issue#34-Create-User-entity-and-configure-Auth

isvyry committed:
Issue#34: Create User entity and configure Auth
Add @Autowired above constructor in the api.mustache template to prevent build failure;
Add user models and endpoints to the spec file;
Add initial SecurityConfig and user entity to configure Auth;
Add initial implementation of Authentication Provider;
Add UserRepository for accessing users in db.

isvyry committed:
Issue#34: Create User entity and configure Auth
Moved spring-starter-security dependency to general pom
Add password encryption

mykytam committed:
Issue#34: Create User entity and configure Auth
Update entities to use RBAC structure

isvyry committed:
Issue#34: Create User entity and configure Auth
Fix problem

Commit:
wip

Commit:
do not use cookies

isvyry committed:
Issue#34: Create User entity and configure Auth
Removed Authentication provider and set UserDetailsService instead

isvyry committed:
Issue#34: Create User entity and configure Auth
Add locked, expired, credentialsExpired, enabled - boolean parameters to User entity

isvyry committed:
Issue#34: Create User entity and configure Auth
Add dev profile in case we need load initial data and use it for tests

isvyry committed:
Merge remote-tracking branch 'origin/dev' into Issue#34-Create-User-entity-and-configure-Auth-problem

# Conflicts:
#	home-data/src/main/resources/application.properties
#	home-open-api/src/main/resources/yaml/openapi.yaml

isvyry committed:
Issue#34: Create User entity and configure Auth
Add newlines at the end of file

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/config/DataLoader.java:
```diff
@@ -0,0 +1,61 @@
+        this.passwordEncoder = passwordEncoder;
+    }
+
+    public void run(ApplicationArguments args) {
```
Is this method overridden from ApplicationRunner? Please mark with appropriate annotation.

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/config/DataLoader.java:
```diff
@@ -0,0 +1,61 @@
+        user1( "ADMIN", "admin@example.com", "UPDATE_NEWS_PERMISSION", "READ_NEWS_PERMISSION", "CREATE_NEWS_PERMISSION");
+    }
+
+    private void user1(String admin1, String email, String... read_news_permission) {
```
Please pay attention to method naming and argument naming. (CamelCase)

isvyry replied:
yeah, these naming totally incorrect, i'll fix it

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/config/DataLoader.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    private void user1(String admin1, String email, String... read_news_permission) {
+        Set<Permission> perm2 = Arrays.stream(read_news_permission).map(perm -> Permission.builder().name(perm).build()).collect(Collectors.toSet());
```
variable naming

isvyry committed:
Issue#34: Create User entity and configure Auth
Fix variable naming

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/config/HomeUserDetailsService.java:
```diff
@@ -0,0 +1,25 @@
+
+@Component
+public class HomeUserDetailsService implements UserDetailsService {
+
```
remove extra new lines

Prosperro-de replied:
some text

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/config/HomeUserDetailsService.java:
```diff
@@ -0,0 +1,25 @@
+    }
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
```
You've declared `throws UsernameNotFoundException`. Have you used it? Please be sure that you meet the contract.

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/config/HomeUserPrincipal.java:
```diff
@@ -0,0 +1,64 @@
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
```
Remove extra newlines

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/config/HomeUserPrincipal.java:
```diff
@@ -0,0 +1,64 @@
+import java.util.Set;
+
+
+public class HomeUserPrincipal implements UserDetails {
```
Please highlight in the class name that it is actually a wrapper for the actual User entity. It would be nice to provide some java doc.

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/config/HomeUserPrincipal.java:
```diff
@@ -0,0 +1,64 @@
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        Set<GrantedAuthority> authorities = new HashSet<>();
+        for ( Role role: user.getRoles()){
```
remove extra spaces

kh0ma replied:
can we do it via java stream?

kh0ma commented on home-application/src/main/resources/config/application.properties:
```diff
@@ -0,0 +1 @@
+spring.profiles.active=dev
```
I don't think it should be active by default

kh0ma commented on home-data/src/main/java/com/softserveinc/ita/homeproject/model/Permission.java:
```diff
@@ -0,0 +1,34 @@
+    private Long id;
+
+    @ManyToMany(mappedBy = "permissions")
+    @EqualsAndHashCode.Exclude
```
We don't have anything to exclude here. Also if we have a primary key it is the only candidate for equals and hashcode.

kh0ma commented on home-data/src/main/java/com/softserveinc/ita/homeproject/model/Role.java:
```diff
@@ -0,0 +1,43 @@
+package com.softserveinc.ita.homeproject.model;
```
Why do we name the package - model? Can we name it an entity? Please don't forget also add module affiliation to the package naming.

kh0ma commented on home-data/src/main/java/com/softserveinc/ita/homeproject/model/Role.java:
```diff
@@ -0,0 +1,43 @@
+import lombok.Setter;
+import lombok.ToString;
+
+import javax.persistence.*;
```
Please don't use import via *. It is a "holivar" actually but please refer to https://stackoverflow.com/a/14897485

kh0ma commented on home-data/src/main/java/com/softserveinc/ita/homeproject/model/Role.java:
```diff
@@ -0,0 +1,43 @@
+    @ManyToMany(fetch = FetchType.EAGER)
+    @EqualsAndHashCode.Exclude
+    @JoinTable(name = "role_permission",
+            joinColumns = @JoinColumn(name = "role_id"),
```
Line identations?

kh0ma commented on home-data/src/main/java/com/softserveinc/ita/homeproject/model/User.java:
```diff
@@ -0,0 +1,60 @@
+    @Column(name = "password")
+    private String password;
+
+    @ManyToMany(fetch = FetchType.EAGER)
```
Why do we need Eager here?

kh0ma commented on home-data/src/main/java/com/softserveinc/ita/homeproject/repository/PermissionRepository.java:
```diff
@@ -0,0 +1,9 @@
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface PermissionRepository extends CrudRepository<Permission, Long> {
```
Just to be sure that you understand what `CrudRepository` does. Do we use everything from it? Wouldn't it be enough to use something smaller?

kh0ma commented on home-server-generator/pom.xml:
```diff
@@ -86,6 +86,7 @@
                                 <parentArtifactId>${project.parent.artifactId}</parentArtifactId>
                                 <parentGroupId>${project.parent.groupId}</parentGroupId>
                                 <parentVersion>${project.parent.version}</parentVersion>
+                                <useBeanValidation>false</useBeanValidation>
```
I don't think we need to disable it but configure properly. Also, I don't think it should be included in these efforts.

kh0ma commented on home-server-generator/src/main/resources/mustaches/server/api.mustache:
```diff
@@ -37,6 +37,7 @@ public class {{classname}}  {
 
     private final {{classname}}Service delegate;
 
+    @Autowired
```
Why do we need it?

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/services/NewsService.java:
```diff
@@ -0,0 +1,51 @@
+
+@Service
+@Profile("dev")
+public class NewsService implements NewsApiService {
```
This one isn't a candidate to be a production code nor dev one. Please write tests and use it only in test scope.

kh0ma commented on pom.xml:
```diff
@@ -43,6 +43,10 @@
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-security</artifactId>
+        </dependency>
```
Please cover by unit tests.

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/config/DataLoader.java:
```diff
@@ -0,0 +1,65 @@
+package com.softserveinc.ita.homeproject.config;
```
Please add the module name identity to the package.

com.softserveinc.ita.homeproject.application.config;

isvyry committed:
Issue#34: Create User entity and configure Auth
Fix package naming
Rewrite getAuthorities() method in a stream way
Delete test service class
Disable dev-profile

isvyry committed:
Issue#34: Create User Entity and configure Auth
Delete Eager FetchType from entity
Fix LazyInitializationException in User Details

isvyry committed:
Issue#34: Create User Entity and configure Auth
Add newline at the end of file

mykytam reviewed (approved):
NICE! GOOD JOB! IHOR TOP!

isvyry committed:
Merge remote-tracking branch 'origin/dev' into Issue#34-Create-User-entity-and-configure-Auth-problem

# Conflicts:
#	home-data/src/main/resources/application.properties
